### PR TITLE
OboeTester: Add channel masks to data path tests

### LINKED ISSUE #1400: OboeTester: add DATA PATH test for down-mixing
Verify proper operation of positional channel masks and down-mixing.
Then transfer to CTS-V.

### PR DESCRIPTION
**Pull request by robertwu1**
Fixes #1400

### CODE REVIEW

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java [hunk 1]
```diff
@@ -478,20 +499,32 @@ void testInputDevices() throws InterruptedException {
...
-                                testInputDeviceCombo(id, deviceType, numChannels, channel);
+                                testInputDeviceCombo(id, deviceType, numChannels, channel, false);
                             }
                         }
                     }
                 }
+                int[] channelMasks = deviceInfo.getChannelMasks();
+                if (channelMasks.length > 0) {
+                    for (int channelMask : channelMasks) {
+                        int nativeChannelMask = channelMask >> 2;
```
Comment (philburk) [inline apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java:519]: Why shift? This looks odd.

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java [hunk 2]
```diff
@@ -521,11 +554,13 @@ void logOneLineSummary(TestResult testResult, String extra) {
 
     void testOutputDeviceCombo(int deviceId,
                                int deviceType,
-                               int numOutputChannels,
+                               int outputChannels,
```
Comment (philburk) [inline apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java:557]: outputChannelCount

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java [hunk 3]
```diff
@@ -538,13 +573,15 @@ void testOutputDeviceCombo(int deviceId,
             int result = testResult.result;
             if (result == TEST_RESULT_FAILED) {
                 if (deviceType == AudioDeviceInfo.TYPE_BUILTIN_EARPIECE
-                        && numOutputChannels == 2
-                        && outputChannel == 1) {
+                        && outputChannels == 2
```
Comment (philburk) [inline apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java:576]: What if this contains a mask?

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java [hunk 4]
```diff
@@ -591,20 +637,32 @@ void testOutputDevices() throws InterruptedException {
...
-                                testOutputDeviceCombo(id, deviceType, numChannels, channel);
+                                testOutputDeviceCombo(id, deviceType, numChannels, channel, false);
                             }
                         }
                     }
                 }
+                int[] channelMasks = deviceInfo.getChannelMasks();
+                if (channelMasks.length > 0) {
+                    for (int channelMask : channelMasks) {
+                        int nativeChannelMask = channelMask >> 2;
```
Comment (philburk) [inline apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java:657]: shift?!

**Review by philburk — CHANGES_REQUESTED** [review]

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java (hunk 3 above)
Comment (robertwu1) [reply apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java]: I changed the function to always take in a count. We can just check that

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java (hunk 4 above)
Comment (robertwu1) [reply apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java:658]: Yes, this is really bizarre but intentional. I added a comment.

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java (hunk 1 above)
Comment (robertwu1) [reply apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java:520]: Yes, this is really bizarre but intentional. I added a comment.

**Comment by philburk** [comment]
Please add a method to convert Java to Oboe Channel Masks.
Hide the shift by 2.

**Comment by robertwu1** [comment]
> Please add a method to convert Java to Oboe Channel Masks. Hide the shift by 2.

Done!

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java [hunk 5]
```diff
@@ -590,23 +595,22 @@ void testOutputDeviceCombo(int deviceId,
...
-                               int outputChannels,
-                               int outputChannel,
-                               boolean useChannelMasks) throws InterruptedException {
+                               int channelCount,
+                               int channelMask,
+                               int outputChannel) throws InterruptedException {
         String typeString = AudioDeviceInfoConverter.typeToString(deviceType);
-        if (useChannelMasks) {
-            int channelCount = Integer.bitCount(outputChannels);
+        if (channelMask != 0) {
```
Comment (philburk) [inline apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java:602]: The code to set the test name should be in the method starting at line 562.

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java [hunk 6]
```diff
@@ -637,29 +641,29 @@ void testOutputDevices() throws InterruptedException {
...
                     for (int channelMask : channelMasks) {
-                        int nativeChannelMask = channelMask >> 2;
-                        // Test higher channel counts.
+                        int nativeChannelMask =
+                                convertJavaChannelMaskToNativeChannelMask(channelMask);
                         log("nativeChannelMask = " + nativeChannelMask + "\n");
                         int channelCount = Integer.bitCount(nativeChannelMask);
                         for (int channel = 0; channel < channelCount; channel++) {
-                            testOutputDeviceCombo(id, deviceType, nativeChannelMask, channel, true);
+                            testOutputDeviceCombo(id, deviceType, nativeChannelMask, 0, channel);
```
Comment (philburk) [inline apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java:666]: The parameters seem out of order. Maybe:
    
    testOutputDeviceCombo(id, deviceType, 0, nativeChannelMask, channel);

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java [hunk 7]
```diff
@@ -465,23 +465,22 @@ void testInputDeviceCombo(int deviceId,
...
     void testInputDeviceCombo(int deviceId,
                               int deviceType,
-                              int inputChannels,
-                              int inputChannel,
-                              boolean useChannelMasks) throws InterruptedException {
+                              int channelCount,
+                              int channelMask,
+                              int inputChannel) throws InterruptedException {
 
         String typeString = AudioDeviceInfoConverter.typeToString(deviceType);
```
Comment (philburk) [inline apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestDataPathsActivity.java:472]: Test naming could move to line 451

**Review by philburk — CHANGES_REQUESTED** [review]

**Review by philburk — APPROVED** [review]

**Pull request merged**